### ISSUE TITLE
Inconsistent behavior when asserting invocations

### ISSUE THREAD

**Issue by embano1** [temporalio/sdk-go#642]
## Expected Behavior

Invocation count assertions provided by the mocking library, such as `Once()` or `Times(n)` show different error/panic behavior affecting test results.

Either it should be documented that these should not be used or fixed, e.g. when recovering from panic and failing the test instead of passing.

## Actual Behavior

This activity is configured with retries. Under test, it throws `PanicError` in the logs when the code is executed multiple times (retries) but the **test succeeds** with `exit code 0`.

```go
env.OnActivity("MyActivity", any, any).Return(nil, errors.New("test failure")).Once() 
```

This activity is also configured with retries but since no error is returned it is only invoked once. Under test, it **fails the test** with a mocking error indicating that the function was only called once, i.e. `exit code != 0`.

```go
env.OnActivity("MyActivity", any, any).Return(nil, nil).Twice()
```

## Steps to Reproduce the Problem

See above

## Specifications

  - Version: `go.temporal.io/sdk v1.10.0`
  - Platform: n/a

**Comment by alfa-alex** [comment]
Ran into the same issue (`go.temporal.io/sdk v1.15.0`).

What's your current position on this? Should the use of invocation count assertions be avoided at the moment?

**Comment by alfa-alex** [comment]
Actually for my use-case there was a relatively simple fix to this: Calling `Test` on the `mock`.

So, instead of the original
```go
// NewTestWorkflowEnvironment creates a new instance of TestWorkflowEnvironment. Use the returned TestWorkflowEnvironment
// to run your workflow in the test environment.
func (s *WorkflowTestSuite) NewTestWorkflowEnvironment() *TestWorkflowEnvironment {
	return &TestWorkflowEnvironment{impl: newTestWorkflowEnvironmentImpl(s, nil)}
}
```
I'm now constructing the `TestWorkflowEnvironment` using:
```go
// NewTestWorkflowEnvironmentWithTest creates a new instance of TestWorkflowEnvironment. Use the returned TestWorkflowEnvironment
// to run your workflow in the test environment.
func (s *WorkflowTestSuite) NewTestWorkflowEnvironmentWithTest(t testing.TB) *TestWorkflowEnvironment {
	env := &TestWorkflowEnvironment{impl: newTestWorkflowEnvironmentImpl(s, nil)}
	env.mock.Test(t)
	return env
}
```
in the `workflow_testsuite.go`. This does actually fail the test in the above case.

Well possible this has undesired side-effects, but maybe something to consider. An alternative to having a different constructor would of course be to add a `Setter` func to the `TestWorkflowEnvironment`, but it felt quite naturally to me to construct the env with the test.

**Comment by alfa-alex** [comment]
So you imply we should check the invocation count assertions via `env.GetWorkflowError()`? I thought that's what `env.AssertExpectations(t)` is for. The documentation reads:
```go
// AssertExpectations  asserts that everything specified with OnActivity
// in fact called as expected.  Calls may have occurred in any order.
```

To give you a minimal working example, change the `Test_Workflow` test of the `helloworld` sample in the `temporal-samples` project to:
```go
	testSuite := &testsuite.WorkflowTestSuite{}
	env := testSuite.NewTestWorkflowEnvironment()

	// Mock activity implementation
	env.OnActivity(Activity, mock.Anything, "Temporal").Return("", errors.New("some error")).Once()
	env.ExecuteWorkflow(Workflow, "Temporal")

	require.True(t, env.IsWorkflowCompleted())
	env.AssertExpectations(t)
```
Even though the output correctly indicates that `Activity` was called more than once, the test doesn't fail. And `env.AssertExpectations` returns true.

Am I misunderstanding the description of that function?

**Comment by cretz** [comment]
> So you imply we should check the invocation count assertions via env.GetWorkflowError()? I thought that's what env.AssertExpectations(t) is for.

My mistake, yes, you can and should use that function if you want to assert the mocking expectations and you are in a test setting. I had forgotten about this helper.

The docs for https://pkg.go.dev/github.com/stretchr/testify/mock#Mock.AssertExpectations (which is what we essentially call) seem to state that it asserts what was called and returned regardless of call order. I wonder if that call does not in fact test invocation count. https://github.com/stretchr/testify/blob/v1.8.0/mock/mock.go#L610-L612 seems to suggest that there would be a failure here (`Repeatability` should be `1` for `Once()`). I will write a test to confirm.

**Comment by alfa-alex** [comment]
Yeah, it looks like a testify issue. I think it's been described here: https://github.com/stretchr/testify/issues/608.

Thank you!

**Comment by cretz** [comment]
> invocation count assertions should probably best be avoided

Concur. And on a general best-practices note, the temptation to overly concern yourself with counts can lead to unnecessary expectations about the internals of the code under test. But if you had to manually count if you needed to confirm invocation counts, you'd be more willing to only do it where it was important to verify behavior.
